python/dataset.cpp

```diff
@@ -281,20 +281,41 @@ void init_dataset(py::module &m) {
                                                         R"(
         View for Dataset, representing a sliced view onto a Dataset;
         Mostly equivalent to Dataset, see there for details.)");
-  datasetView.def(py::init<Dataset &>());
+  py::options options;
+  options.disable_function_signatures();
+  datasetView.def(py::init<Dataset &>(), py::arg("dataset"),
+                  R"(__init__(dataset: Dataset) -> None
+                    Initialises from viewed Dataset.
+                    )");
 
   py::class_<Dataset> dataset(m, "Dataset", R"(
-    Dict of data arrays with aligned dimensions.)");
+  Dict of data arrays with aligned dimensions.)");
 
-  dataset.def(py::init<const std::map<std::string, DataArrayConstView> &>())
-      .def(py::init<const DataArrayConstView &>())
+  dataset
       .def(py::init([](const std::map<std::string, VariableConstView> &data,
                        const std::map<Dim, VariableConstView> &coords) {
              return Dataset(data, coords);
            }),
            py::arg("data") = std::map<std::string, VariableConstView>{},
-           py::arg("coords") = std::map<Dim, VariableConstView>{})
-      .def(py::init([](const DatasetView &other) { return Dataset{other}; }))
+           py::arg("coords") = std::map<Dim, VariableConstView>{},
+           R"(Overloaded Function
+             
+             __init__(self, data: Dict[str, VariableConstView] = {}, coords: Dict[str, VariableConstView] = {})
+          
+             Initialise from a dictionary of Variables and optionally coordinates.
+             Each Variable is converted into a DataArray within the Dataset.
+             )")
+      .def(py::init<const std::map<std::string, DataArrayConstView> &>(),
+           py::arg("data"),
+           R"(Overloaded Function
+           
+              __init__(self, data: Dict[str, DataArrayConstView])
+              
+              Initialise from a dictionary of DataArrays.
+              )");
```

Comment:
http://xarray.pydata.org/en/stable/generated/xarray.Dataset.html accepts both variables and data arrays in the same constructor. Can we do that?

Reply:
Yes probably can I'll have a look into it. Two questions:
* Do we want to allow a dict made up of a mix of DataArrays and Variables or restrict them to only one.
* How do we want to handle the case where the user passed a dict of DataArrays and a dict of coords? This is currently not an allowed overload but could it be?

Reply:
In pricple the mix is ok, I think. Maybe this constructor can be written on the Python side and simply use something like:
```python
ds = sc.Dataset()
for coord in coords:
 ds.coord[coord] = ...
for data in data:
 ds[data] = ...
```
Then your two questions would be handled automatically?